### ISSUE TITLE
Reading FoF halos hdf5 structure

### ISSUE THREAD

**Issue by jegarfa** [Kambrian/HBT2#15]
Hi, I'm new in HBTplus.
I have generated FoF halos as first step in using HBTplus, and would like to read them in Python. I noticed that the toolbox provides routines to read Subhalos but not FoF halos.

When using h5py I can see that there are two keys in the HDF5 file: ['HaloParticles', 'SnapshotId']. However, I am having troubles reading the rest of the HDF5 file structure, specifically in retrieving the basic properties of halos such as positions and masses. Since you have a better understanding of the file structure, could you please provide an example of how to read these basic properties of the hdf5 snapshot (e.g. positions, masses from "HaloSnap.hdf5")?
Thanks!

**Comment by Kambrian** [comment]
Hi, you do not read those properties because they actually do not exist there. 

As you have found the halo files contain only the list of particles for each halo. You can find the FoF mass of the halo by counting the number of particles and multiply by particle mass. If want the center of mass you will have to match the particle ids with the particles in the simulation and compute the average position.

However, if you further process the halos by HBT+ to identify subhalos, you can then use the position of the central subhalo in each FoF as a more robust proxy for halo center. 

After that, if you further run halo_virial in toolbox with the subhalo catalog, you can get a HaloSize folder containing the virial sizes of each halo, which is again a more formal proxy for the mass of a halo.

**Comment by jegarfa** [comment]
great! thanks a lot for the clarification.

**Comment by jegarfa** [comment]
Hi, I followed your advice, however I found some strange results. 

I am using the groups from hbt+ and matching the particle ids with the particles in the simulation. However I found somthing odd when I plot the groups of particles. These are the results when I plot all the particles from the groups:
![image](https://github.com/Kambrian/HBTplus/assets/10395230/15dc00cb-8b54-44c6-abe8-5fab664a59b5)

the result is even more strange when I plot the individual groups, for example if I choose only particles from the first group, I expected to have a bunch of particles close each other as a single group, however I got this:

![image](https://github.com/Kambrian/HBTplus/assets/10395230/ff051f24-eccd-450b-a2c5-475117727b18)
 
![image](https://github.com/Kambrian/HBTplus/assets/10395230/933a4979-9594-4caa-853a-1da4f9aaeaf5)

Have you seen this behaviour before? for some reason the groups does not seem to follow the DM distribution of the simulation. I did several tests changing the softening lenght and the minimun number of particles to consider a halo, however the result was almost the same. Do you have any idea what might be causing this behavior? any advice, suggestion is welcome. If helpful, the snapshot can be found here (http://gg.gg/15vjym). This is the parameterfile I am using:

```
[Compulsary Params]
SnapshotPath  snap_0.000
HaloPath output
SubhaloPath output
SnapshotFileBase gadgetz0.000
MaxSnapshotIndex 7                  
BoxSize 1000.0                          
SofteningHalo 0.4                     

[Reader]
SnapshotFormat gadget 

GroupFileFormat hbt_group_hdf5 
MaxConcurrentIO 4  

[FoF]
FoFLinkParam 0.2      
FoFDarkMatterMassFraction 0.846 #OmegaDM0/OmegaMatter0
FoFSizeMin 30         

[Units]
MassInMsunh 1e10
LengthInMpch 1
VelInKmS 1

[Performance]
ParticleIdNeedHash 0 
PeriodicBoundaryOn 1
```

**Comment by Kambrian** [comment]
Incorrect fof output usually means the snapshot was not correctly loaded when running fof or hbt. I suggest you to first check if the units have been correctly specified. 

To help further verify if the snapshot has been loaded correctly, you can use `src/io/snapshot_io.test` in the hbt source directory. To do so, please first update your code to the most recent version (I've just pushed an update to fix the makefile there). Then
```
cd src/io
make snapshot_io.test -B
./snapshot_io.test ../../myconfig.conf 0
```
This will print some information about how hbt has been interpreting your snapshot. You can easily modify the main function at the end of `src/io/snapshot_io.cpp` to print more information for diagnosis.

**Comment by jegarfa** [comment]
Hi, thanks a lot for the update. 

I checked it out and the simulation is being read well. I got all the expected values (I added a few more from the header) and the particles, also the units are the correct ones. The issue was when I read the output and sorted the inidices. There was a missmatch between the index of the output and the ids of the simulation. 

```
Reading configuration file ../../configs/Example-minisim.conf
 ( 8 total files ) : 16777216 particles loaded.
ParticleMass:525.222
Hz :100
ScaleFactor :1
OmegaM0 :0.3175
OmegaLambda0 :0.682407
Number Of Particles:16777216
Id of 10th particle: 2097141
position: (118.244, 994.142, 952.413, )
Mass: 525.222, Mass of 100th particle:525.222
Index of 10th particle : 10
```

this is the new and correct result:

![image](https://github.com/Kambrian/HBTplus/assets/10395230/9994157a-e96f-4fb7-b5ab-837186669892)

![image](https://github.com/Kambrian/HBTplus/assets/10395230/f6f210c6-11c3-4fe7-a71b-205ffb0f44d4)

thanks!
